### PR TITLE
Hello World

### PR DESCRIPTION
**Pull request by marnen**
Hello World and some basic track configuration. A few things I still need help on:

* There's no README. This is because `configlet generate .` doesn't work: it runs, but generates no files, even when I have the problem-specifications project checked out next to the track.
* Where should the global test library go? In `/global`?

### CODE REVIEW

**Comment by iHiD** [comment]
What a crazy language!! This is going to be a fun track! :)

**Comment by marnen** [comment]
@kytrinx:

> This pull request contains three different types of changes.
> I'd prefer to have three separate pull requests, one for each general topic.

I could do that (and I normally do!), but for something this small, it seemed like more trouble than it was worth...

> The exercism thing is that you'll need to run `configlet generate . --only hello-world` in the hello-world branch in order to get the README.

As I said in my initial comment, that didn't work.

**Comment by kytrinyx** [comment]
> As I said in my initial comment, that didn't work.

Sorry, I totally missed that. 

> for something this small, it seemed like more trouble than it was worth

For me it really isn't about size, but about atomicity. It makes it a lot harder to find things later when they've been mixed together (I particularly have a lot of anxiety around this because I work in several hundred repositories at any one time).

Like I said, though, totally happy to do the legwork on splitting it up, since (a) it's pretty low effort for me, and (b) it's my anxiety.

**Comment by kytrinyx** [comment]
> Where should the global test library go? In /global?

Yepp, in `global`.

> configlet generate . doesn't work: it runs, but generates no files, even when I have the problem-specifications project checked out next to the track.

I'll debug this tomorrow!

**Comment by marnen** [comment]
@kytrinyx:
> For me it really isn't about size, but about atomicity. 

Oh, I totally get that. I usually find, though, that on *any* project of mine, the first pull request or two isn't terribly atomic, because I usually do a lot of configuration only as I figure out that it's necessary—i.e., while I'm doing something that depends on that configuration.

Regardless, though, I can break this up.

> It makes it a lot harder to find things later when they've been mixed together

Out of curiosity, doesn't `git blame` work better for that anyway? (I'm a great believer in a clear and immutable branch history, but not so much for that reason as to keep different lines of development separate, which explains my confusion. But maybe that's what you meant. :) )

**Comment by marnen** [comment]
@kytrinyx:

One more question.

>>Where should the global test library go? In `/global`?

> Yepp, in `global`

...but it gets delivered to the root directory of the track, right? If so, that will mess up my relative include paths and make running tests harder, unless I can come up with something clever (which perhaps I can).

**Comment by kytrinyx** [comment]
> Out of curiosity, doesn't git blame work better for that anyway?

Kind of. A lot of times I can't even remember which repo a discussion happened in, so I'm searching for keywords on GitHub to try to find the issue or PR so that I can eventually find the code and track down history. In this case that's not really relevant, because all of the discussions are completely related to this track.

This is all really just my own survival mechanisms. I have the complete opposite philosophy with respect to git history. I don't care about how it happened originally. I almost always discover how I want to do something as I'm messing with things, and branch off of a fresh master to do that in a separate, isolated place, then rebase to work on top of that change. I totally get it if you hate that! I see the value of both approaches, I've just ended up in the one camp due to personal preferences and circumstance, I think.

> ...but it gets delivered to the root directory of the track, right?

Yepp.

If so, that will mess up my relative include paths and make running tests harder, unless I can come up with something clever (which perhaps I can).

Hm. Yeah, I can definitely see that. What a lot of tracks do is copy all the files to a temporary directory and run it from there. That's generally because the test suites have skips in them, though, and rewriting files in place so that you can run the tests is a _great_ way to accidentally commit changes.

But a temp dir might actually work in this case as well, even if there's no skips or whatever to delete.

**Comment by marnen** [comment]
@kytrinyx:

>  I have the complete opposite philosophy with respect to git history. I don't care about how it happened originally. I almost always discover how I want to do something as I'm messing with things, and branch off of a fresh master to do that in a separate, isolated place, then rebase to work on top of that change. I totally get it if you hate that!

I do hate that, very much. :) I've come to believe that Git history should generally reflect what actually happened, and that rebasing and rewriting history beyond an occasional `commit --amend` are not good ideas in the context of using Git, because once you use them, Git is no longer a VCS (because a VCS has to maintain an immutable history of _something_, yada yada yada). But that's a discussion for another venue; suffice it to say that the "sane subset" of Git that I use well over 99% of the time does not include any variety of `git rebase`, and that I've come to this view in part because every time I try to rewrite history, the commit tree becomes _harder_ to manipulate as a direct result. See http://paul.stadig.name/2010/12/thou-shalt-not-lie-git-rebase-ammend.html if you care. :)

Basically, every commit (except for occasional temporary commits that are never pushed from my own machine) is guaranteed to represent a state that I've tested and verified to be consistent. If I rewrite history, I lose the ability to make that guarantee, and there's no way to tell that I've done so.

As a matter of interest, though, I was planning on separating out the various pull requests by cherry-picking separate topical commits, as long as I verify each cherry-picked commit to make sure that it is consistent in its new home.

**Comment by marnen** [comment]
@kytrinyx:

Back onto discussing more immediately useful things...

> > If so, that will mess up my relative include paths and make running tests harder, unless I can come up with something clever (which perhaps I can).

> Hm. Yeah, I can definitely see that. What a lot of tracks do is copy all the files to a temporary directory and run it from there.

I think you may have misunderstood. Of course I can have Travis do that (though in this case, I don't think I need to, because of the way this track's tests are constructed); anything can be automated (I'll probably write some custom Rake tasks or shell scripts). The problem is how to run tests locally as I develop my example solution in a test-first way. All the test utility functions are `)COPY`ed in (basically, included) from `test.apl`, but that's `/global/test.apl` in the repo and `/test.apl` in the tracks that will run on the user's machine.

Hmm. Maybe I need a dev Rake task to run tests from a temp directory in the way that a user would (and rename the example file so the test runner picks it up too). Yuck.

> That's generally because the test suites have skips in them, though, and rewriting files in place so that you can run the tests is a great way to accidentally commit changes.

I haven't bothered to use skips in the APL tests so far. I considered it, but there doesn't seem to be any reason to do so.

**Comment by kytrinyx** [comment]
I pushed up a fix (to this branch) for the README generation. I named the template wrong so it wasn't getting picked up :/

Since this is your track, and you have a very well-reasoned approach to why you're doing it that way, I'm completely content with letting you drive the rules and norms around how to approach git history and pull requests. In other words: merge at will, with or without cherry-picking into separate branches.

> See http://paul.stadig.name/2010/12/thou-shalt-not-lie-git-rebase-ammend.html if you care. :)

Thanks, I've read it, and understand and respect the point of view. I completely agree... for master, though I would not use language like _sin_ or _lie_ when talking about this, because I don't think it has anything to do with ethics or dishonesty, and using that language seems unnecessarily polarizing [EDIT: in retrospect I suspect that the author might be doing the tongue-in-cheek thing. I tend to not pick up on that]. I have come to change my mind about stuff like this before, when I come to understand the trade-offs differently.

For me, making master immutable pragmatic rather than moral. Rewriting master has many potentially painful and potentially disastrous consequences in terms of collaboration, and I don't think the trade-offs are worth it. I have done it on occasion when I accidentally have committed massive binaries and I don't want to impose the cost of downloading those onto future contributors. When I do, I do it with trepidation, much communication with existing contributors, and much hand-wringing.

When it comes to branches that I am working on, I currently find that I care more about communicating my understanding of the changes—which develops and changes along the way—than how I got there. _/shrug_. Trade-offs. Ask me again in five years, I suppose.

> Back onto discussing more immediately useful things...

Hah. Yes. I have ideas, give me a few minutes to think this through.

**Comment by kytrinyx** [comment]
> I think you may have misunderstood.

Yes! I think I did.

Since no tracks have implemented the global thing yet, and I'm mid-rewrite on the new CLI, there is absolutely no reason why we can't hash out the details until we find something that feels elegant.

My immediate thought is: stick the files anywhere you like, and have a section in the `config.json` that lists the file paths of the files that should be downloaded.

That way people could put them wherever it makes sense for their language—either in root, or tucked away in a directory, and nobody would have to care about some agreed-upon convention, including the CLI. It would be explicit in the file list in the config.

**Comment by marnen** [comment]
@kytrinyx:

> I pushed up a fix (to this branch) for the README generation. I named the template wrong so it wasn't getting picked up :/

Yes, I saw. Thanks!

> Since this is your track, and you have a very well-reasoned approach to why you're doing it that way, I'm completely content with letting you drive the rules and norms around how to approach git history and pull requests. In other words: merge at will, with or without cherry-picking into separate branches.

I'll probably still separate this out as you requested, both because you requested it, and because I _do_ prefer more atomic pull requests when possible, and kind of felt like I was being a little sloppy with making these less so.


> >See http://paul.stadig.name/2010/12/thou-shalt-not-lie-git-rebase-ammend.html if you care. :)
> Thanks, I've read it, and understand and respect the point of view. I completely agree... for master, though I would not use language like sin or lie when talking about this, because I don't think it has anything to do with ethics or dishonesty, and using that language seems unnecessarily polarizing [EDIT: in retrospect I suspect that the author might be doing the tongue-in-cheek thing. I tend to not pick up on that].

I probably wouldn't have used those terms either, though I do sometimes adopt terminology like Stadig's when discussing this issue. :)

> I have come to change my mind about stuff like this before, when I come to understand the trade-offs differently.

As have I, and I too have done things differently in the past...and it was painful, so I don't do that anymore. :)

> When it comes to branches that I am working on, I currently find that I care more about communicating my understanding of the changes—which develops and changes along the way—than how I got there.

I've heard a number of people make this same argument, but for me it does not hold water. My usual response is this: As a maintainer, I want to see what actually happened, not what you thought might look good to someone else. In other words, when I examine a Git history, I think of myself kind of as a historian: I'm mostly interested in the most accurate records possible, not in the "spin" that the original developer thought I'd want to see. To me that's of suspect accuracy, and therefore is to be disregarded. (This is actually also the same reason I don't trust or pay much attention to comments in code...)

> /shrug. Trade-offs. Ask me again in five years, I suppose.

:) And likewise, assuming I'm still using Git. (I like Git a lot, but debates like this point to some cases that Linus and Junio did not consider. For example, I'd enjoy having a VCS that immutably version-controlled the state of the tree, so that one _could_ rewrite history without forever losing the old version.)

**Comment by marnen** [comment]
@kytrinyx:

> Since no tracks have implemented the global thing yet, and I'm mid-rewrite on the new CLI, there is absolutely no reason why we can't hash out the details until we find something that feels elegant.

> My immediate thought is: stick the files anywhere you like, and have a section in the `config.json` that lists the file paths of the files that should be downloaded.

Hmm, yeah. I'll play around with that.

**Comment by kytrinyx** [comment]
> I'd enjoy having a VCS that immutably version-controlled the state of the tree, so that one could rewrite history without forever losing the old version.

Me too!

> I'll play around with that.

Perfect, if you have any insights or thoughts along the way, feel free to discuss. I'd love to get this right. Or right-er.

**Comment by marnen** [comment]
I'm thinking of some sort of `setup_files` property in the config file, with an array of file names to be copied to the track by the `setup` command. As for the names, I'm thinking that shell glob syntax would be most convenient, but regexes would be more powerful and consistent with `test_pattern`. Thoughts?

**Comment by kytrinyx** [comment]
> As for the names, I'm thinking that shell glob syntax would be most convenient, but regexes would be more powerful and consistent with test_pattern. Thoughts?

I have a slight preference for regex for the sake of consistency, but if you find a reason to prefer the shell glob I'm open to that too.

**Comment by marnen** [comment]
@kytrinyx The main reason for preferring shell globs is the ugly double escape needed for periods in the "JSON-string-as-regex" syntax we're using. I agree that having some globs and some regexes is not ideal, though.

**Comment by kytrinyx** [comment]
How would you feel about using a character group for periods? `[.]`?

**Comment by marnen** [comment]
I saw that syntax elsewhere. It's still 3 characters for 1, and I need to double-check that `.` isn't a meta character within brackets.

**Comment by marnen** [comment]
Closing and will break up.

- [ ] General track config stuff (e.g. test pattern, git ignore)
- [ ] Add yourself to the maintainer file
- [ ] Hello World
